### ISSUE TITLE
Issue with coordinator, lagging  for seconds  

### ISSUE THREAD

**Issue by mikeatm** [arangodb/arangodb#2324]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ x] 3.0
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [x] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

With  both  3.1.11-1 and  3.2.alpha2-1,  im getting an unexpected slow response from the coordinator both with
arangosh and over the web UI,  it takes  several seconds to move from any of the menu items (from cluster to nodes or to  collections and so on). Here are logs from the coordinator in  loglevel debug 

```
2017-02-23T16:29:52Z [29224] DEBUG connecting to ip endpoint 'http+tcp://172.16.6.23:8529'
2017-02-23T16:29:52Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:29:52Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:29:52Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:02Z [29224] WARNING cannot contact coordinator CRDN-ecf883d2-8df4-42e4-9c5e-f52f92b1b8fa on endpoint tcp://172.16.6.23:8529
2017-02-23T16:30:02Z [29224] DEBUG connecting to ip endpoint 'http+tcp://172.16.6.23:8529'
2017-02-23T16:30:02Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:02Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:02Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] WARNING cannot contact coordinator CRDN-ecf883d2-8df4-42e4-9c5e-f52f92b1b8fa on endpoint tcp://172.16.6.23:8529
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG connecting to ip endpoint 'http+tcp://172.16.6.23:8529'
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:12Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:23Z [29224] WARNING cannot contact coordinator CRDN-ecf883d2-8df4-42e4-9c5e-f52f92b1b8fa on endpoint tcp://172.16.6.23:8529
2017-02-23T16:30:23Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:23Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:23Z [29224] DEBUG waiting for unused V8 context
2017-02-23T16:30:23Z [29224] DEBUG connecting to ip endpoint 'http+tcp://172.16.6.23:8529'
```

I stopped the coordinator and run it with log level trace.  

```
2017-02-23T17:12:48Z [31639] TRACE leaving V8 context 0
2017-02-23T17:12:48Z [31639] TRACE returned dirty V8 context
2017-02-23T17:12:48Z [31639] TRACE V8 context has reached GC timeout threshold and will be scheduled for GC
2017-02-23T17:12:48Z [31639] TRACE found unused V8 context
2017-02-23T17:12:48Z [31639] TRACE entering V8 context 0
2017-02-23T17:12:48Z [31639] TRACE Request: POST /_api/agency/read HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflat
e\r\ncontent-type: application/json\r\nContent-Length: 31\r\n\r\n[["arango/Supervision/Health"]]
2017-02-23T17:12:48Z [31639] DEBUG waiting for unused V8 context
2017-02-23T17:12:48Z [31639] DEBUG waiting for unused V8 context
2017-02-23T17:12:48Z [31639] DEBUG waiting for unused V8 context
2017-02-23T17:12:48Z [31639] TRACE leaving V8 context 0
2017-02-23T17:12:48Z [31639] TRACE returned dirty V8 context
2017-02-23T17:12:48Z [31639] TRACE V8 context has reached GC timeout threshold and will be scheduled for GC
2017-02-23T17:12:48Z [31639] TRACE found unused V8 context
2017-02-23T17:12:48Z [31639] TRACE entering V8 context 0
2017-02-23T17:12:48Z [31639] DEBUG waiting for unused V8 context
2017-02-23T17:12:48Z [31639] DEBUG waiting for unused V8 context
2017-02-23T17:12:48Z [31639] DEBUG waiting for unused V8 context
2017-02-23T17:12:48Z [31639] TRACE downloading file. endpoint: tcp://172.16.6.23:8529, relative URL: http://172.16.6.23:8529/_db/_system/_admin/aardvark/statistics/short?cou
nt=1
2017-02-23T17:12:48Z [31639] TRACE Request: GET /_db/_system/_admin/aardvark/statistics/short?count=1 HTTP/1.1\r\nHost: 172.16.6.23\r\nConnection: Keep-Alive\r\nUser-Agent: 
ArangoDB\r\n\r\n
2017-02-23T17:12:48Z [31639] DEBUG connecting to ip endpoint 'http+tcp://172.16.6.23:8529'
2017-02-23T17:12:48Z [31639] TRACE bind to address '172.16.6.23', port 8529
2017-02-23T17:12:48Z [31639] TRACE Request: POST /_api/agency/transient HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflate\r\ncontent-type: application/json\r\nContent-Length: 144\r\n\r\n[[{"arango/Sync/ServerStates/CRDN-d527ec60-14f2-423e-a107-442b11223047":{"new":{"status":"SERVING","time":"2017-02-23T17:12:48Z"},"op":"set"}}]]
2017-02-23T17:12:48Z [31639] TRACE Request: POST /_api/agency/read HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflate\r\ncontent-type: application/json\r\nContent-Length: 267\r\n\r\n[["arango/Current/Version","arango/Current/Foxxmaster","arango/Current/FoxxmasterQueueupdate","arango/Plan/Version","arango/Shutdown","arango/Sync/Commands/CRDN-d527ec60-14f2-423e-a107-442b11223047","arango/Sync/UserVersion","arango/Target/FailedServers","/.agency"]]
2017-02-23T17:12:48Z [31639] TRACE Cannot find an agency persisted in RAFT 8|
2017-02-23T17:12:49Z [31639] TRACE Request: POST /_api/agency/transient HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflate\r\ncontent-type: application/json\r\nContent-Length: 144\r\n\r\n[[{"arango/Sync/ServerStates/CRDN-d527ec60-14f2-423e-a107-442b11223047":{"new":{"status":"SERVING","time":"2017-02-23T17:12:49Z"},"op":"set"}}]]
2017-02-23T17:12:49Z [31639] TRACE Request: POST /_api/agency/read HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflate\r\ncontent-type: application/json\r\nContent-Length: 267\r\n\r\n[["arango/Current/Version","arango/Current/Foxxmaster","arango/Current/FoxxmasterQueueupdate","arango/Plan/Version","arango/Shutdown","arango/Sync/Commands/CRDN-d527ec60-14f2-423e-a107-442b11223047","arango/Sync/UserVersion","arango/Target/FailedServers","/.agency"]]
2017-02-23T17:12:49Z [31639] TRACE Cannot find an agency persisted in RAFT 8|
2017-02-23T17:12:50Z [31639] TRACE Request: POST /_api/agency/transient HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflate\r\ncontent-type: application/json\r\nContent-Length: 144\r\n\r\n[[{"arango/Sync/ServerStates/CRDN-d527ec60-14f2-423e-a107-442b11223047":{"new":{"status":"SERVING","time":"2017-02-23T17:12:50Z"},"op":"set"}}]]
2017-02-23T17:12:50Z [31639] TRACE Request: POST /_api/agency/read HTTP/1.1\r\nHost: 172.16.6.19\r\nConnection: Keep-Alive\r\nUser-Agent: ArangoDB\r\nAccept-Encoding: deflate\r\ncontent-type: application/json\r\nContent-Length: 267\r\n\r\n[["arango/Current/Version","arango/Current/Foxxmaster","arango/Current/FoxxmasterQueueupdate","arango/Plan/Version","arango/Shutdown","arango/Sync/Commands/CRDN-d527ec60-14f2-423e-a107-442b11223047","arango/Sync/UserVersion","arango/Target/FailedServers","/.agency"]]
2017-02-23T17:12:50Z [31639] TRACE Cannot find an agency persisted in RAFT 8|
```

seems like there is an issue with:
1. JS contexts, not sure whether this is normal, lots of log lines  with waiting for context
2. Raft errors?  lots of lines about  not finding agency persisted in RAFT. 

And there is this NaN i dont understand. 
![myshot](https://cloud.githubusercontent.com/assets/8317804/23268661/b5da9928-f9ee-11e6-82d7-7983d5c42047.jpg)

I would appreciate some advice. This is fedora 25.  I was running a cluster previously with an earlier 3x release without incident.   Here are some commands used to run the cluster (all on 172.16.6.x network), 

Coordinator

```
/usr/sbin/arangod  --log.level trace  --log.output file:///var/log/arangodb3/primary.log  --server.endpoint tcp://172.16.6.23:8529  --cluster.my-address  tcp://172.16.6.23:8529  --server.authentication false     --cluster.agency-endpoint tcp://172.16.6.15:5001  --cluster.agency-endpoint tcp://172.16.6.19:5001  --cluster.agency-endpoint tcp://172.16.6.23:5001 --cluster.my-local-info coord1 --cluster.my-role COORDINATOR --database.directory coordinator
```

One of the primary:

```
/usr/sbin/arangod  --log.level  debug  --log.output file:///var/log/arangodb3/primary.log  --server.endpoint tcp://172.16.6.23:8529  --cluster.my-address  tcp://172.16.6.19:8529  --server.authentication false     --cluster.agency-endpoint tcp://172.16.6.15:5001  --cluster.agency-endpoint tcp://172.16.6.19:5001  --cluster.agency-endpoint tcp://172.16.6.23:5001 --cluster.my-local-info db1 --cluster.my-role PRIMARY --database.directory  db1 
```

one of the  agency:

```
/usr/sbin/arangod --log.level debug --log.output file:///var/log/arangodb3/agency.log --server.endpoint tcp://172.16.6.23:5001 --agency.my-address=tcp://172.16.6.23:5001  --server.authentication false --agency.activate true  --agency.size 3    --agency.endpoint tcp://172.16.6.15:5001  --agency.endpoint tcp://172.16.6.19:5001  --agency.endpoint tcp://172.16.6.23:5001 --agency.supervision true --database.directory   agency2
```

There is no indication of errors in the primary and the agency at debug level  logging.

**Comment by mikeatm** [comment]
Here is some number to give a feeling of how slow it is, this is the same command in the arangsh tutorial on the _system database in an empty cluster,

```
172.16.6.23:8529@_system> console.time("dbsave");

172.16.6.23:8529@_system> db._create('place2s');
[ArangoCollection 4113808, "place2s" (type document, status loaded)]

172.16.6.23:8529@_system> console.timeEnd("dbsave");
2017-02-24T14:37:53Z [17844] INFO dbsave: 45408ms
```

something is wrong, pointers on what to investigate? 
here are some logs from one agency and a log level trace from a coordinator
[primary.log-20170224.gz](https://github.com/arangodb/arangodb/files/799617/primary.log-20170224.gz)

[agency.log-20170224.gz](https://github.com/arangodb/arangodb/files/799622/agency.log-20170224.gz)

**Comment by mikeatm** [comment]
Seems this is a network related issue,  if  i bind arangodb to  a different interface it works in cluster mode, 
but it  slows down to a crawl for  tun* interface. 
Does anyone know whether theirs a known issue with arangodb and tun interfaces (an overlay network with openvpn)?  @dothebart  close this issue if this is known, but if someone can explain this, it would be fantastic.

**Comment by mikeatm** [comment]
@kvahed  i did try with   both 3.1.11-1 and 3.2.alpha2-1,  (see the first post),  and ended up trying 2.8x, 3.0x, 3.1.0 -> 3.1.11  but they were all not working. At this point i strongly suspected my environment, but could not be sure. 
I recently changed  network architecture, and no longer  used the physical  ethx device, but an overlay network on  tunX  that provides some privacy using openvpn.  on changing back to ethX device,  everything worked  and t was no longer  taking 45 seconds to run  `db._create('place2s');`   and  the UI  is as responsive as before the migration to tunX devices.  Did not need to go though  tun, but tried to, did not work. 
Hardware:
RAM 1g,  1 core.   (minimal i guess).

tl;dr
- We solved the extreme slow response.
- problem was caused by binding to tunX device. 
- we tried  2.8, 3.0x, 3.1.0x -> 3.1.11x  none worked bound to tunx.
- binding to ethX  slow response disappears.

**Comment by mikeatm** [comment]
False positive, the problem is  still present,  not solved, 
is this because of the machine specs?  because in non cluster mode it runs fine,
but when clustered it behaves like this.

**Comment by kvahed** [comment]
do you have some details for me please: #machines, #CPU, #RAM, Data size for arangodb? as i had mentioned earlier there is no reasonable way that tunX by itself creates a problem

**Comment by mikeatm** [comment]
Details:
3 machines, each with:
```
1 Core: Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz
RAM:    MemTotal:        1016524 kB
```
All machines are identical,  with:
Machine 1:
agency1, primary1
Machine 2
agency2,  primary 2
Machine 3
agency3,  coordinator

Data Size,
I probably failed to make this clear,  its  `empty`  this is a freshly deployed  cluster instance.
I tore down the infrastructure and rebuilt  everything  over the weekend,  with identical results, 
OS: Fedora 25, 
Kernel: 4.9.9-200.fc25.x86_64

Arangodb  versions tried:
2.8  and 3.0x,  3.1 -> 3.1.12 ,  3.2-alpha-1,  negative results on all. 
On a cloud provider  (DO) i was able to run 2.8 in cluster without a hitch. so there is
something wrong, but i cant tell what.

**Comment by mikeatm** [comment]
has anyone tested arangodb with IP aliasing?  im unable to find any other difference between the working and not-working  environments.

**Comment by mikeatm** [comment]
We can safely close this, seems it was the number of threads,  
i changed this to

    agency:4
    primary:4
    coordinator: 8
The default  (0 for all) seems to be the root cause of  my issue. 
Is it possible to add this to the documentation for those that setup the cluster using the process method?
